File: src/quests/timeutil.py
"""Deadline / duration helpers. DB and API wire format: UTC."""

from __future__ import annotations

from datetime import datetime, timedelta, timezone

UTC = timezone.utc


def ensure_utc(dt: datetime | None) -> datetime | None:
    """Treat naive datetimes as UTC (SQLite round-trip); convert aware → UTC."""
    if dt is None:
        return None
    if dt.tzinfo is None:
        return dt.replace(tzinfo=UTC)
    return dt.astimezone(UTC)


def to_db_utc(dt: datetime | None) -> datetime | None:
    """UTC wall time without tzinfo — what SQLite DateTime stores/returns."""
    dt = ensure_utc(dt)
    if dt is None:
        return None
    return dt.replace(tzinfo=None)


def to_utc_iso(dt: datetime | None) -> str | None:
    """RFC3339 UTC with trailing Z — safe for JS Date()."""
    dt = ensure_utc(dt)
    if dt is None:
        return None
    text = dt.isoformat(timespec="microseconds")
    # +00:00 → Z for clearer UTC signal to browsers
    if text.endswith("+00:00"):
        return text[:-6] + "Z"
    return text

# Default urgent window when deadline is set without an explicit duration.
DEFAULT_DURATION_SECONDS = 24 * 60 * 60


def auto_duration_seconds(deadline_at: datetime, anchor: datetime) -> int:
    """If duration omitted: 24h (or time left to deadline if shorter), at least 60s."""
    deadline_at = ensure_utc(deadline_at)
    anchor = ensure_utc(anchor)
    assert deadline_at is not None and anchor is not None
    left = int((deadline_at - anchor).total_seconds())
    if left <= 60:
        return 60
    return min(DEFAULT_DURATION_SECONDS, left)


def window_start(deadline_at: datetime, duration_seconds: int) -> datetime:
    deadline_at = ensure_utc(deadline_at)
    assert deadline_at is not None
    return deadline_at - timedelta(seconds=max(1, duration_seconds))


def is_in_urgent_window(
    deadline_at: datetime | None,
    duration_seconds: int | None,
    *,
    now: datetime | None = None,
) -> bool:
    """True when (deadline − duration) < now (urgent window has started)."""
    if deadline_at is None or not duration_seconds:
        return False
    now = ensure_utc(now or datetime.now(UTC))
    deadline_at = ensure_utc(deadline_at)
    assert now is not None and deadline_at is not None
    return window_start(deadline_at, duration_seconds) < now


def remaining_seconds(
    deadline_at: datetime | None,
    *,
    now: datetime | None = None,
) -> int | None:
    if deadline_at is None:
        return None
    now = ensure_utc(now or datetime.now(UTC))
    deadline_at = ensure_utc(deadline_at)
    assert now is not None and deadline_at is not None
    return int((deadline_at - now).total_seconds())


def timer_tone(
    deadline_at: datetime | None,
    duration_seconds: int | None,
    *,
    now: datetime | None = None,
) -> str | None:
    """green | orange | red — by remaining / duration fractions (2/3, 1/3)."""
    if deadline_at is None or not duration_seconds:
        return None
    rem = remaining_seconds(deadline_at, now=now)
    if rem is None:
        return None
    dur = max(1, int(duration_seconds))
    frac = rem / dur
    if frac > 2 / 3:
        return "green"
    if frac > 1 / 3:
        return "orange"
    return "red"


def format_remaining(seconds: int | None) -> str:
    if seconds is None:
        return ""
    sign = "-" if seconds < 0 else ""
    s = abs(int(seconds))
    h, rem = divmod(s, 3600)
    m, sec = divmod(rem, 60)
    if h:
        return f"{sign}{h}:{m:02d}:{sec:02d}"
    return f"{sign}{m}:{sec:02d}"

File: go/internal/timeutil/timeutil.go
package timeutil

import (
	"strings"
	"time"
)

const DefaultDurationSeconds = 24 * 60 * 60

var layouts = []string{
	time.RFC3339Nano,
	time.RFC3339,
	"2006-01-02T15:04:05.999999",
	"2006-01-02T15:04:05",
	"2006-01-02 15:04:05.999999",
	"2006-01-02 15:04:05",
}

func NowUTC() time.Time {
	return time.Now().UTC()
}

func EnsureUTC(t time.Time) time.Time {
	if t.IsZero() {
		return t
	}
	if t.Location() == time.UTC {
		return t
	}
	return t.UTC()
}

// ToDBUTC formats for SQLite naive UTC storage (matches SQLAlchemy DateTime).
func ToDBUTC(t time.Time) string {
	t = EnsureUTC(t)
	return t.Format("2006-01-02 15:04:05.000000")
}

func ToUTCISO(t *time.Time) *string {
	if t == nil || t.IsZero() {
		return nil
	}
	tt := EnsureUTC(*t)
	s := tt.Format("2006-01-02T15:04:05.000000Z")
	return &s
}

func ParseFlexible(s string) (time.Time, error) {
	s = strings.TrimSpace(s)
	if s == "" {
		return time.Time{}, nil
	}
	if strings.HasSuffix(s, "Z") {
		if t, err := time.Parse(time.RFC3339Nano, s); err == nil {
			return t.UTC(), nil
		}
		if t, err := time.Parse(time.RFC3339, s); err == nil {
			return t.UTC(), nil
		}
	}
	for _, layout := range layouts {
		if t, err := time.ParseInLocation(layout, s, time.UTC); err == nil {
			return t.UTC(), nil
		}
	}
	return time.Time{}, &ParseError{Value: s}
}

type ParseError struct{ Value string }

func (e *ParseError) Error() string { return "invalid datetime: " + e.Value }

func AutoDurationSeconds(deadline, anchor time.Time) int {
	left := int(deadline.Sub(anchor).Seconds())
	if left <= 60 {
		return 60
	}
	if left < DefaultDurationSeconds {
		return left
	}
	return DefaultDurationSeconds
}

func NormalizeDeadline(deadline *time.Time, duration *int, durationExplicit bool, now time.Time) (*time.Time, *int) {
	if deadline == nil {
		return nil, nil
	}
	d := EnsureUTC(*deadline)
	if durationExplicit && duration != nil {
		v := *duration
		if v <= 0 {
			// Explicit 0 (or negative) means "no window": deadline stays as a
			// plain target time, no urgent-window timer, no forced auto-expire.
			zero := 0
			return &d, &zero
		}
		return &d, &v
	}
	sec := AutoDurationSeconds(d, EnsureUTC(now))
	return &d, &sec
}

func RemainingSeconds(deadline *time.Time, now time.Time) *int {
	if deadline == nil {
		return nil
	}
	rem := int(EnsureUTC(*deadline).Sub(EnsureUTC(now)).Seconds())
	return &rem
}

func WindowStart(deadline time.Time, durationSeconds int) time.Time {
	if durationSeconds < 1 {
		durationSeconds = 1
	}
	return EnsureUTC(deadline).Add(-time.Duration(durationSeconds) * time.Second)
}

func IsInUrgentWindow(deadline *time.Time, duration *int, now time.Time) bool {
	if deadline == nil || duration == nil || *duration <= 0 {
		return false
	}
	return WindowStart(*deadline, *duration).Before(EnsureUTC(now))
}

func TimerTone(deadline *time.Time, duration *int, now time.Time) *string {
	if deadline == nil || duration == nil || *duration <= 0 {
		return nil
	}
	rem := RemainingSeconds(deadline, now)
	if rem == nil {
		return nil
	}
	dur := *duration
	if dur < 1 {
		dur = 1
	}
	frac := float64(*rem) / float64(dur)
	var tone string
	switch {
	case frac > 2.0/3.0:
		tone = "green"
	case frac > 1.0/3.0:
		tone = "orange"
	default:
		tone = "red"
	}
	return &tone
}
